Dont call each when calling body on response to fix #23964

rthbound opened the pull request:
- Adds `#each_chunk` to ActionDispatch::Response. It's a method which will be called by `ActionDispatch::Response#each`.
- Make `Response#each` a proper method instead of delegating to `@stream`
- In Live, instead of overriding `#each`, override `#each_chunk`.
- `#each` should just spit out `@str_body` if it's already set.
- Adds `#test_set_header_after_read_body_during_action` to prove this fixes #23964.
- Adds `#test_each_isnt_called_if_str_body_is_written` to ensure `#each` is not called when `@str_body` is available.
- Call `@response.sent!` in AC::TestCase's `#perform` so a test response acts a bit more like a real response. This makes tests that call  `#assert_stream_closed` pass again.
- Additionally, assert `#committed?` in `#assert_stream_closed`.
- Make test that was calling `@response.stream.each` pass again by calling `@response.each` instead.

rthbound committed:
Fixes #23964

  - Adds #each_chunk to ActionDispatch::Response. it's a method which
    will be called by ActionDispatch::Response#each.
  - Make Response#each a proper method instead of delegating to @stream
  - In Live, instead of overriding #each, override #each_chunk.
  - `#each` should just spit out @str_body if it's already set
  - Adds #test_set_header_after_read_body_during_action
    to prove this fixes #23964
  - Adds #test_each_isnt_called_if_str_body_is_written to
    ensure #each_chunk is not called when @str_body is available
  - Call `@response.sent!` in AC::TestCase's #perform so a test response
    acts a bit more like a real response. Makes test that call  `#assert_stream_closed`
    pass again.
  - Additionally assert `#committed?` in `#assert_stream_closed`
  - Make test that was calling @response.stream.each pass again by
    calling @response.each instead.

sgrif commented on actionpack/lib/action_dispatch/http/response.rb:
```diff
@@ -96,11 +96,11 @@ def initialize(response, buf)
-                        each { |chunk| buf << chunk }
-                        buf
-                      end
+        begin
```
We can't remove the memoization here, this makes the `body` method potentially very expensive, and would break streaming responses.

rthbound replied:
@sgrif tests passed locally with the memoization added back. Travis should pick that change up momentarily.

sgrif commented:
After looking at this a bit more, I'm not entirely comfortable with this approach. The `each` method is defined that way for a reason, and I suspect this could interact poorly with streaming responses. I'd prefer that we work to separate out code required for streaming from the buffer.

rthbound commented:
@sgrif are you sure? The only change here is that calling `body` won't call `each`. `each` should be unchanged, and I don't see where `response.body` is used internally (except in tests).

sgrif commented:
The issue is that you can now leave the response in an invalid state simply by calling `body` on it.

rthbound commented:
@sgrif I beleive that's why we got rid of the memoization that was used when calling #body (which rails is not calling internally, except in tests).

sgrif commented:
You're misunderstanding me. You cannot call `#each` more than once if the response is streaming. The implementation of `#each` is written to ensure that right now. With this change, a user calling `response.body` ever with a streaming response will leave things in an invalid state.

rthbound commented:
@sgrif okay, thanks for that. I do understand you now. I checked with @matthewd. He might soon have some suggestions for something "similar-but-still-different" to resolve the issue.

rthbound commented:
@sgrif can you think of a way to show with a test that a user calling `response.body` for a streaming response in an application or a gem would result in some breakage with these changes? i.e. the test this PR should have broken.

jeremy commented:
Is this blocking Rails 5.0.0 release candidate? Is it blocking beta4?

sgrif commented:
Blocking RC, yes. Not blocking beta 4

On Sun, Apr 24, 2016, 3:12 PM Jeremy Daer <notifications@github.com> wrote:

> Is this blocking Rails 5.0.0 release candidate? Is it blocking beta4?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/24029#issuecomment-214038126>
>

rthbound commented:
@sgrif I'm under the impression that my latest has addressed your concerns.

Also, I did add the tests I mentioned.

Just trying to make I'm on the same page as everyone else.

sgrif commented:
I will review later this week, thanks

On Sun, Apr 24, 2016, 8:24 PM Ryan T. Hosford <notifications@github.com>
wrote:

> @sgrif <https://github.com/sgrif> I'm under the impression that my latest
> has addressed your concerns.
>
> Also, I did add the tests I mentioned.
>
> Just trying to make I'm on the same page as everyone else.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/24029#issuecomment-214095869>
>

sgrif commented on actionpack/lib/action_dispatch/http/response.rb:
```diff
@@ -129,6 +138,12 @@ def close
+      private
+
+        def each_chunk(&block)
+          @buf.each(&block) # extract into own method
```
What is this comment for?

sgrif commented on actionpack/test/dispatch/live_response_test.rb:
```diff
@@ -65,7 +65,7 @@ def test_headers_cannot_be_written_after_webserver_reads
         latch = Concurrent::CountDownLatch.new
 
         t = Thread.new {
-          @response.stream.each do
```
I think this behavior needs to remain unchanged. In a streaming response, we should only be able to call each on the stream once.

sgrif commented on actionpack/lib/action_controller/metal/live.rb:
```diff
@@ -210,6 +202,14 @@ def on_error(&block)
+      private
+
+        def each_chunk(&block)
+          while str = @buf.pop
```
I think we need to retain the `sending!` and `sent!` calls here. Streaming responses are the only place that we shouldn't be able to call `each` multiple times.

sgrif commented:
OK after a second review I think I have a better understanding. Thanks for working on this

rthbound commented:
Thank you @sgrif, and also thanks to @matthewd for working closely with me on this.

sgrif commented:
:heart:
